### DISCUSSION TITLE
Greenwood init command and templates

### DISCUSSION THREAD

**Discussion by thescientist13** [ProjectEvergreen/greenwood#770] [Ideas]
## Overview

It would be great if there was an easy way for users to initialize a Greenwood project and have all the scaffolding provided including
- _package.json_ with `@greenwood/cli` pre-installed
- _greenwood.config.js_
- _src/_ directory with an initial page

```
├── greenwood.config.js
├── .gitignore
├── package.json
└── src/
             └─ pages/
                        └─ index.md
```

Additionally, GitHub repos could be used as templates to easily provide a ready made starting point that could easily be shared with users and not dependent on being coded / configured / bundled directly into the `init` command itself.

## Proposal
Create a new `init` package in the monorepo that can be invoked via npm / `npx` and that supports scaffolding as well as pulling from a predefined (or custom) template

Some examples
```js
// just spits out the bare minimum scaffolding
$ npx @greenwood/init

// use a predefined Greenwood template, which would just be its own GitHub repo
$ npx @greenwood/init --template=blog

// use a forked or custom Greenwood template, which can be a _users_ own GitHub / GitLab / etc repo
$ npx @greenwood/init --template=github:<username>/<repo-name>
```

Additionally in future iterations, `init` could also ask for / do more things like
- running `npm init`
- asking for name / location of the package
- installing more dependencies / plugins

## Action Items
Let's plan the work out across a few PRs, iterating incrementally following this rough sequence
1. [x] Create new `init` package that can generate a bare bones Greenwood repo - https://github.com/ProjectEvergreen/greenwood/issues/777
    - _greenwood.config.js_
    - _src/pages/index.md_ directory
    - _package.json_
    - _README.md_ (maybe with some nots for how to get started, link to docs
1. [x] Introduce a `--template=xxx` flag that can consume from an existing repo and scaffold it out file for file
    - Let's create at least one repo for this in our org, e.g. **greenwood-template-blog** and expose **blog** as the first template type option - https://github.com/ProjectEvergreen/greenwood/issues/841
1. [ ] Allow someone to prefix the value for `--template` with `github` and provide their own repo location
    - Let's have one of us fork the **greenwood-template-blog** and use that for testing, eg.
      ```sh
      $ npx @greenwood/init --template=github:thescientist13/greenwood-template-blog
      ```
----

From there we can evaluate and prioritize other features and enhancements
  - Support installs from GitLab or Bitbucket repos
  - Prompt for additional configurations with the default `init` command
  - Ability to install plugins via `init`

**Comment by hutchgrant** [comment]
should we include the ability to create the project directory or are we under the assumption they've already created a project directory ?

**Reply by thescientist13** [reply]
I think it's safe to assume that we are operating in the `cwd`.  I think that's what most init tools do I assume, but might be good to do a quick survey to see if there is an established convention.  

Perhaps later on we could add a --dir flag so someone could do both at once?  e.g.
```sh
$ npx @greenwood/init --dir ./my-new-app
```

Not sure but I think it's safe to assume `.` for now, at least to start.

cc: @aholtzman

**Comment by hutchgrant** [comment]
Rather than copy a default package.json over as we did with `create-evergreen-app` should we instead execute `npm init` and `npm i @greenwood/cli` ?

**Reply by hutchgrant** [reply]
Another issue with this is we need to modify the package.json to have the `greenwood build` and `develop` scripts etc.

**Reply by thescientist13** [reply]
Hmm, running `npm init` for them is something I hadn't considered but is an interesting idea.  I think most tools assume you will run it.  Typically they just scaffold things out then let you take over.  Maybe another opportunity to cross reference the DX workflows of other similar tools.

Though it does raise a good question about what the init contents of _package.json_ should look like.  I know to _publish_ you need things like `name` and `version` (and I assume some others) though for the sake of development I don't think anything prevents us from just spitting out something like this
```json
{
  "scripts": {
    "start": "greenwood develop",
    "build": "greenwood build",
    "serve": "greenwood server"
  },
  "devDependencies": {
    "@greenwood/cli": "~0.y.z"
   }
}
```

Though I'm sure we could easily stub out the `name` from the current folder name or eventually ask the user as a feature?

I think if at minimum we can spit something out that allows them to `npm i && npm start` then we're in a good place so that we can get into a package and add some tests, then we can enrich that experience / _package.json_ and basically just put in as much work as time allows.  🙌

**Comment by hutchgrant** [comment]
What kind of folder structure should we be expecting for a template.  Take the [getting-started](https://github.com/ProjectEvergreen/greenwood-getting-started) repo for example, perhaps not all those files are needing to be scaffolded.  Maybe only the required files should be placed in `template` directory ? Or we just assume everything in a template repo should be scaffolded.  How is that any different than just doing a git clone ?  I suppose if you already have some pages, you could run @greenwood/init --template blog and it would copy over just the needed files rather than replace.

**Reply by thescientist13** [reply]
>  Maybe only the required files should be placed in template directory ? Or we just assume everything in a template repo should be scaffolded. _How is that any different than just doing a git clone_ ?

Yeah, I was thinking at least the initial thing to try was just grab from the repo directly and then just only filter through the files we would actually care about, like
- _greenwood.config.js_
- _package.json_
- _src/_

You could also look into using [**degit**](https://github.com/Rich-Harris/degit).

Perhaps if someone does want _all_ the files in the repos, that could be done by adding another flag?
```sh
$ npx @greenwood/cli --template=blog --all
```
----

As a note, I wouldn't get too hung up on the implementation details, just because it is as easy as a `git clone` to the grab the contents, doesn't diminish the value of also then preparing the right files and getting things installed and setup all in one command.  Once you the nail the outcome, you can always refactor under the hood over time.

**Comment by hutchgrant** [comment]
I also think it would be a good idea to curate some "suggested" templates that one can select from a list in the CLI

**Reply by thescientist13** [reply]
Yup, I think that's covered in the section action item and that we would start initially with a blog template.  I can make the repo if we need; _greenwood-template-blog_.

**Comment by thescientist13** [comment]
Just calling out we should probably really flesh out the current blog starter, as it doesn't really "work" per se.  For example, none of the links work.  Now that we have refreshed the website, perhaps we can also refresh the styles of the starter as well
https://github.com/ProjectEvergreen/greenwood-template-blog

**Comment by thescientist13** [comment]
I think now that we have a good implementation around this we can move more specifically to what kind of template / project options we want to commit to and build out for the `init` command, so have created a dedicated discussion for this - https://github.com/ProjectEvergreen/greenwood/discussions/1335
